Mesh modification and displacements improvements for UOs and mortar

GiudGiud opened the pull request:
related to @vanwdani 's work

GiudGiud committed:
Reset the range for boundary nodes and elements when freeing them
refs #31067

GiudGiud committed:
Improve checking & comment in nearest node value aux

GiudGiud committed:
Re-init more with user objects on the displaced mesh
Instead of limiting to elem, also do face and neighbors as needed

GiudGiud committed:
Add some logging to sidset updater
_ too easy to get the logic wrong and delete the sideset while trying to move it

GiudGiud committed:
Re-init elements, faces and neighbors for internal side and interface UOs on the displaced mesh
This makes logical sense. I don't have a need for it right now since the UO I am working on is a DomainUO

GiudGiud committed:
Update mortar functors from nonlienar system once mortar mesh has changed

This is only important if there are new subdomains added, and mortar should be active on them

GiudGiud committed:
Add a test for mortar + displaced + mesh modification

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/3092937/), step Generate coverage on d898ac1 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>bbdd03</th>
      <th colspan=4>#31105 d898ac</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.54%</td>
      <td>85.55%</td>
      <td>+0.01%</td>
      <td>73.33%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>116127</td>
      <td>116157</td>
      <td>+30</td>
      <td>22</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>19626</td>
      <td>19622</td>
      <td>-4</td>
      <td>8</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/31105/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/31105/coverage/framework)


## Modules coverage

Inconsistent report tags were found between the head and base reports.
This can happen when reports are missing from either the head or the base.

Inconsistent tags:
  module-chemical_reactions-restep
  module-combined-restep
  module-contact-restep
  module-electromagnetics-restep
  module-external_petsc_solver-restep
  module-fluid_properties-restep
  module-fsi-restep
  module-functional_expansion_tools-restep
  module-geochemistry-restep
  module-heat_transfer-restep
  module-level_set-restep
  module-misc-restep
  module-navier_stokes-restep
  module-optimization-restep
  module-peridynamics-restep
  module-phase_field-restep
  module-porous_flow-restep
  module-ray_tracing-restep
  module-rdg-restep
  module-reactor-restep
  module-richards-restep
  module-scalar_transport-restep
  module-solid_mechanics-restep
  module-solid_properties-restep
  module-stochastic_tools-restep
  module-subchannel-restep
  module-thermal_hydraulics-restep
  module-xfem-restep

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/31105/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/31105/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/31105/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/31105/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/31105/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/31105/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/31105/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/31105/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/31105/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/31105/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/31105/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/31105/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/31105/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/31105/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/31105/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/31105/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/31105/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/31105/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/31105/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/31105/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/31105/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/31105/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/31105/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/31105/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/31105/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/31105/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/31105/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/31105/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/31105/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 73.33% is less than the suggested 90.0%



This comment will be updated on new commits.

lindsayad commented on framework/include/systems/NonlinearSystemBase.h:
```diff
@@ -70,6 +70,8 @@ class NonlinearSystemBase : public SolverSystem, public PerfGraphInterface
 
   virtual void preInit() override;
+  /// Update the mortar functors if the mesh has changed
+  virtual void reinitMortarFunctors();
```
```suggestion
  void reinitMortarFunctors();
```

lindsayad commented on framework/src/auxkernels/NearestNodeValueAux.C:
```diff
@@ -35,12 +35,15 @@ NearestNodeValueAux::NearestNodeValueAux(const InputParameters & parameters)
-    mooseError("NearestNodeValueAux can only be used with one boundary at a time!");
+    paramError("boundary", "NearestNodeValueAux can only be used with one boundary at a time!");
+  if (boundaryNames().empty())
+    paramError("boundary", "Should be specified on a boundary or group of boundaries");
```
If we're erroring if there's more than one boundary, seems like we should remove the second half of this sentence
```suggestion
    paramError("boundary", "Should be specified on a boundary");
```

lindsayad commented on framework/src/meshmodifiers/SidesetAroundSubdomainUpdater.C:
```diff
@@ -165,6 +166,15 @@ SidesetAroundSubdomainUpdater::finalize()
+  {
+    if (_remove.size())
+      _console << "Removing " << _remove.size() << " sides from sideset " << _boundary_id
+               << " on processor " << _pid << std::endl;
+    if (_add.size())
+      _console << "Adding " << _add.size() << " sides to sideset " << _boundary_id
+               << " on processor " << _pid << std::endl;
```
I don't have a great suggestion for what to change this to, but the current language makes it seem like the boundary information will be out of sync across processes, e.g. we're only adding/removing it on `_pid` but in truth we're going to do a global communication to make sure we do it in sync across processes. Honestly might be best just to remove the part including the `_pid`

lindsayad commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -4315,12 +4315,14 @@ FEProblemBase::addUserObject(const std::string & user_object_name,
+      // Whether to re-init or not depends on the attributes of the base classes.
+      // For example, InterfaceUOBase has "_current_side_elem" and "_neighbor_elem"
+      // so it needs to reinit on displaced faces and neighbors
+      if (euo || nuo || duo || isuo || iuob)
         _reinit_displaced_elem = true;
```
`_reinit_displaced_elem` is supposed to correspond to reinitialization on bulk/volumetric element quadrature points. `_reinit_displaced_face` should be for the `_current_side_elem` and I would expect `_reinit_displaced_neighbor` to be for `_neighbor_elem` unless `_neighbor_elem` corresponds to a bulk neighbor element and quadrature points

lindsayad replied:
If there's going to be no differentiation between element interior and element face, then we should just delete the `_reinit_displaced_face` data member

GiudGiud replied:
the bug occured with the domain UO. If you look at this initial logic
```
      if (euo || nuo || duo)
        _reinit_displaced_elem = true;
      else if (suo || duo)
        // shouldn't we add isuo
        _reinit_displaced_face = true;
      else if (iuob || duo)
        _reinit_displaced_neighbor = true;
```
you see that domain UO catches the first 'if' so when you use the displaced _current_neighbor you get garbage. 

Now the current logic
```
if (euo || nuo || duo || isuo || iuob)
        _reinit_displaced_elem = true;
if (suo || duo || isuo || iuob)
        _reinit_displaced_face = true;
if (iuob || duo || isuo)
        _reinit_displaced_neighbor = true;
```

euo, nuo or duo : all good (now for duo, not previously)
isuo / iuo : we are setting the reinit elem for nothing (since included in face and neighbor). Does not hurt though
We are reinitting the face because that s what you act on.
We are reinitting the neighbor because the interface / internal side concept could include the neighbor
suo: we should be 1 sided.

GiudGiud replied:
so basically:
_reinit_displaced_elem -> _current_elem is up to date
_reinit_displaced_face -> _current_elem and _current_side is up to date
_reinit_displaced_neighbor -> _current_neighbor is up to date

I am not entirely sure it works like this, but that would be a fine way to work imo

GiudGiud replied:
> So the simple presence of a _current_elem in the data members of a class should not be sufficient to trigger the toggling of _reinit_displaced_elem to true

I dont think it hurts though.

GiudGiud replied:
should be fixed. Let s see if diff is related

lindsayad commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -4315,12 +4315,14 @@ FEProblemBase::addUserObject(const std::string & user_object_name,
-      else if (suo || duo)
-        // shouldn't we add isuo
+      if (suo || duo || isuo || iuob)
         _reinit_displaced_face = true;
-      else if (iuob || duo)
+      if (iuob || duo || isuo || iuob)
```
I bet these variable names are Robert's work 😆 I don't like them because they're so short. Would you be willing to at least change `iuob` to `iuo`? The `b` suffix makes me first think of "boundary" instead of "base"

lindsayad commented on test/tests/mortar/mesh_modification/tests:
```diff
@@ -0,0 +1,11 @@
+    input = 'displaced_and_modifying_meshes.i'
+    csvdiff = 'displaced_and_modifying_meshes_out.csv'
+    abs_zero = 1e-8
+    requirement = 'The system shall be able to use mortar constraints while modifying element subdomains in the mesh and the displaced mesh.'
```
```suggestion
    requirement = 'The system shall be able to use mortar constraints while modifying element subdomains in the reference mesh and the displaced mesh.'
```

GiudGiud committed:
Address Alex's review
- interfaceUO -> iuo
- fix test req
- avoid unnecessary virtual
- clear up confusion around the logging for sideset updating

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

GiudGiud committed:
Further adjust displaced mesh reinit logic:
- dont reinit on displaced neighbor or lowerD if only reiniting on elements
- dont reinit on element if only using side user objects

lindsayad reviewed (approved):
Thank you for the fixes. The `else if` was particularly embarrassing/shocking that we had that messed up

GiudGiud commented:
umm. This is tied to the test design. I'll have to find some middle ground. 

I ll do this once I m done with OSTI day
